zcash_client_sqlite/src/wallet/common.rs

```diff
@@ -116,6 +116,163 @@ where
+            params,
+            account,
+            zats,
+            anchor_height,
+            exclude,
+            protocol,
+            to_spendable_note,
+        ),
+    }
+}
+
+/// selects all the spendable notes available above ``MARGINAL_FEE`` amount from a given account
+/// and specified shielded protocols excluding the ones in the exclude slice.
+///
+/// - Implementation details:
+///   - Notes with individual value *below* the ``MARGINAL_FEE`` will be ignored
+///   - Note spendability is determined using the `anchor_height`
+#[allow(clippy::too_many_arguments)]
+fn select_maximum_spendable_notes<P: consensus::Parameters, F, Note>(
+    conn: &Connection,
+    params: &P,
+    account: AccountUuid,
+    anchor_height: BlockHeight,
+    exclude: &[ReceivedNoteId],
+    protocol: ShieldedProtocol,
+    to_spendable_note: F,
+) -> Result<Vec<ReceivedNote<ReceivedNoteId, Note>>, SqliteClientError>
+where
+    F: Fn(&P, &Row) -> Result<Option<ReceivedNote<ReceivedNoteId, Note>>, SqliteClientError>,
+{
+    let birthday_height = match wallet_birthday(conn)? {
+        Some(birthday) => birthday,
+        None => {
+            // the wallet birthday can only be unknown if there are no accounts in the wallet; in
+            // such a case, the wallet has no notes to spend.
+            return Ok(vec![]);
+        }
+    };
+
+    let (table_prefix, index_col, note_reconstruction_cols) = per_protocol_names(protocol);
+    if unscanned_tip_exists(conn, anchor_height, table_prefix)? {
+        return Ok(vec![]);
+    }
+
+    // The goal of this SQL statement is to select all the spendable notes above the
+    // MARGINAL_FEE (5000 zats).
+    // 1) Use a window function to create a view of all notes, ordered from oldest to
+    //    newest
+    //
+    // 2) Select all unspent notes in the desired account.
+    let mut stmt_select_notes = conn.prepare_cached(
+        &format!(
+            "WITH eligible AS (
+                 SELECT
+                     {table_prefix}_received_notes.id AS id, txid, {index_col},
+                     diversifier, value, {note_reconstruction_cols}, commitment_tree_position,
+                     accounts.ufvk as ufvk, recipient_key_scope
+                 FROM {table_prefix}_received_notes
+                 INNER JOIN accounts
+                    ON accounts.id = {table_prefix}_received_notes.account_id
+                 INNER JOIN transactions
+                    ON transactions.id_tx = {table_prefix}_received_notes.tx
+                 WHERE accounts.uuid = :account_uuid
+                 AND {table_prefix}_received_notes.account_id = accounts.id
+                 AND value > 5000 -- FIXME #1316, allow selection of dust inputs
+                 AND accounts.ufvk IS NOT NULL
+                 AND recipient_key_scope IS NOT NULL
+                 AND nf IS NOT NULL
+                 AND commitment_tree_position IS NOT NULL
+                 AND transactions.block <= :anchor_height
```

Comment:
It occurs to me that there's a race condition here: if someone makes a withdrawal from an exchange to the wallet, and then attempts to send max, the maximum balance of the wallet will not be sent but it would still allow the transaction to be constructed; the notes received in the last 10 blocks would simply not contribute to the balance. This is probably not what the user is intending; we should require that all notes apart from dust be outside the reorg window in order to allow sending max.

Reply:
ohhhhhhh nice catch!!! I'll try to build a test that breaks this and work on the fix from it.

Reply:
If I make this fail from within this statement, I won't be able to catch the specific failure will I? I think that this particular failure needs to have its own error so wallets can inform that to the user specifically

Reply:
A reasonable approach would be, for send-max, instead of passing the anchor height in to the query, instead return the minimum possible anchor height _from_ the query. Then the wallet can effectively schedule the transaction for the target height that is appropriate.

Reply:
I added a test that produces this scenario, but I'm struggling with the error types, I think that a Sqlite error is too low level and specific to this implementation of the backend but couldn't manage to juggle generics so that a more clear ProposalError is thrown instead.